sklearn/utils/optimize.py

```diff
@@ -89,7 +118,8 @@ def _cg(fhess_p, fgrad, maxiter, tol):
         Ap = fhess_p(psupi)
         # check curvature
         curv = np.dot(psupi, Ap)
-        if 0 <= curv <= 3 * np.finfo(np.float64).eps:
+        if 0 <= curv <= 16 * np.finfo(np.float64).eps * psupi_norm2:
+            # See https://arxiv.org/abs/1803.02924, Algo 1 Capped Conjugate Gradient.
             break
         elif curv < 0:
             if i > 0:
```

Comment:
Info: Codecov complains about no test coverage of those lines.
We have only one very simple test for `_newton_cg` in `test_optimize.py`. Also the tests for LogisticRegression with newton-cg don't hit these branches.

I don't think blowing up the tests is beneficial for this PR. I could add a TODO note in the code, though.